modules/spring-data-2.2-ext/src/main/java/org/apache/ignite/springdata22/repository/query/IgniteRepositoryQuery.java

```diff
@@ -838,98 +805,59 @@ private Query prepareQuery(IgniteQuery qry, DynamicQueryConfig config, ReturnStr
-        final BinaryObjectBuilder bldr = binary.builder(binType.typeName());
+    private <V> V rowToEntity(List<?> row, FieldsQueryCursor<?> cursor) {
+        Constructor<?> ctor;
 
-        for (int i = 0; i < row.size(); i++) {
-            final GridQueryFieldMetadata fMeta = meta.get(i);
-            final String metaField = fMeta.fieldName();
-            // add existing entity fields to binary object
-            if (binType.field(fMeta.fieldName()) != null && !metaField.equalsIgnoreCase(QueryUtils.KEY_FIELD_NAME)
-                && !metaField.equalsIgnoreCase(QueryUtils.VAL_FIELD_NAME)) {
-                final Object fieldValue = row.get(i);
-                if (fieldValue != null) {
-                    final Class<?> clazz = getClassForBinaryField(binary, binType, fMeta);
-                    // null values must not be set into binary objects
-                    bldr.setField(metaField, fixExpectedType(fieldValue, clazz));
-                }
-            }
-            else {
-                // don't want key or val column... but wants null values
-                if (!metaField.equalsIgnoreCase(QueryUtils.KEY_FIELD_NAME) && !metaField.equalsIgnoreCase(
-                    QueryUtils.VAL_FIELD_NAME))
-                    metadata.put(metaField, row.get(i));
-            }
-        }
-        return bldr.build().deserialize();
-    }
-
-    /**
-     * Obtains real field class from resultset metadata field whether it's available
-     */
-    private Class<?> getClassForBinaryField(final IgniteBinaryImpl binary,
-        final BinaryType binType,
-        final GridQueryFieldMetadata fieldMeta) {
         try {
-            final String fieldId = fieldMeta.schemaName() + "." + fieldMeta.typeName() + "." + fieldMeta.fieldName();
-
-            if (binaryFieldClass.containsKey(fieldId))
-                return binaryFieldClass.get(fieldId);
-
-            Class<?> clazz = null;
-
-            synchronized (binaryFieldClass) {
-
-                if (binaryFieldClass.containsKey(fieldId))
-                    return binaryFieldClass.get(fieldId);
+            ctor = type.getDeclaredConstructor();
 
-                String fieldName = null;
-
-                // search field name on binary type (query returns case insensitive
-                // field name) but BinaryType is not case insensitive
-                for (final String fname : binType.fieldNames()) {
-                    if (fname.equalsIgnoreCase(fieldMeta.fieldName())) {
-                        fieldName = fname;
-                        break;
-                    }
-                }
+            ctor.setAccessible(true);
+        }
+        catch (NoSuchMethodException | SecurityException ignored) {
+            ctor = null;
+        }
 
-                final CacheObjectBinaryProcessorImpl proc = (CacheObjectBinaryProcessorImpl)binary.processor();
+        try {
+            Object res = ctor == null ? GridUnsafe.allocateInstance(type) : ctor.newInstance();
```

Comment:
Can we also support "_VAL" field (perhaps by another ticket)? In case of "_VAL" field presents, we can return its value and don't construct new instances using reflection. It will allow to use entities even if not all fields exposed to SQL.

Reply:
Yes, we can. I will create the corresponding ticket.

Reply:
https://issues.apache.org/jira/browse/IGNITE-13647